### ISSUE TITLE
borg mount fails if mount point is not writable

### ISSUE THREAD

**Issue by Feuermurmel** [borgbackup/borg#4948]
## Have you checked borgbackup docs, FAQ, and open Github issues?

Yes

## Is this a BUG / ISSUE report or a QUESTION?

Issue

#### Your borg version (borg -V).

```
$ borg -V
borg 1.1.9
```

#### Operating system (distribution) and version.

Debian stretch.

#### Hardware / network configuration, and filesystems used.

Local directory on ZFS. Hardware is a crummy multi-slot USB disk enclosure, but ZFS shows 0 I/O errors on the pool.

#### How much data is handled by borg?

2 to 2.5 TiB.

#### Full borg commandline that lead to the problem (leave away excludes and passwords)

```
borg mount /srv/backup/apu.borg::2017-09-17-211548 /srv/backup/borg-migrate
```

#### Can you reproduce the problem? If so, describe how. If not, describe troubleshooting steps you took before opening the issue.

The issue I'm encountering seems to be quite simple. When I try to use `borg mount` to mount an archive I get this error message:

```
/srv/backup/borg-migrate: Mountpoint must be a writable directory
```

The statement in the error message is true. I've taken a habit of setting the _immutable_ flag on mount points, to prevent anything being written there and getting hidden behind the mount. `borg mount` seems to needlessly check whether the directory over which the Fuse filesystem is mounted is writable, even though it never needs to write to that directory _or_ I am completely mistaken and that check is necessary, in which case I suggest adding a quick explanation of this restriction to the documentation.

**Comment by ThomasWaldmann** [comment]
can you mount an ext4 fs into such an immutable mountpoint dir?

can you mount a non-borg fuse fs into such an immutable mountpoint dir?

**Comment by Feuermurmel** [comment]
> can you mount an ext4 fs into such an immutable mountpoint dir?

Yes. I create loopback device, create an ext4 file system, create a mount point, set its immutable bit and mount the file system over it:

```
$ truncate --size=1G test
$ losetup --show --find test
/dev/loop0
$ mke2fs -t ext4 /dev/loop0
mke2fs 1.43.4 (31-Jan-2017)
Discarding device blocks: done                            
Creating filesystem with 262144 4k blocks and 65536 inodes
Filesystem UUID: 99937776-05a5-4416-b748-c113870318e9
Superblock backups stored on blocks: 
	32768, 98304, 163840, 229376

Allocating group tables: done                            
Writing inode tables: done                            
Creating journal (8192 blocks): done
Writing superblocks and filesystem accounting information: done

$ mkdir mountpoint
$ chattr +i mountpoint
$ mkdir mountpoint/test
mkdir: cannot create directory ‘mountpoint/test’: Operation not permitted
$ mount /dev/loop0 mountpoint
$ mkdir mountpoint/test
$ ls mountpoint/
lost+found  test
$ umount mountpoint/
```

> can you mount a non-borg fuse fs into such an immutable mountpoint dir?

Yes. Here I mount an SSHFS mount into the same mount point:

```
$ sshfs localhost:/boot mountpoint/
$ ls mountpoint
System.map-4.9.0-11-amd64  config-4.9.0-9-amd64       initrd.img-4.9.0-9-amd64
System.map-4.9.0-9-amd64   grub			      vmlinuz-4.9.0-11-amd64
config-4.9.0-11-amd64	   initrd.img-4.9.0-11-amd64  vmlinuz-4.9.0-9-amd64
$ umount mountpoint/
```

But I did notice one peculiarity. Mounting a fuse filesystem as a regular user requires the user to have write access to the mount point:

```
$ mkdir mountpoint
$ chmod a-w mountpoint/
$ sshfs localhost:/boot mountpoint/
fusermount: failed to open /etc/fuse.conf: Permission denied
fusermount: user has no write access to mountpoint /home/michi/mountpoint
```

This is documented in the [`mount.fuse(8)`](https://manpages.debian.org/buster/fuse3/mount.fuse.8.en.html) man page:

> The user can only mount on a mountpoint, for which it has write permission

So it seems that the check I called _"needlessly"_ is necessary, unless `borg run` is run as root.

**Comment by ThomasWaldmann** [comment]
I can have a look into our code, but could be that this is implemented by / required by fuse code we just call.

**Comment by infectormp** [comment]
I get same message when try mount repository to non-existent directory 
```
borg mount /backup/directum/sql /tmp/sal_13
/tmp/sal_13: Mountpoint must be a writable directory
```
`/tmp/sal_13` is not exist

**Comment by ThomasWaldmann** [comment]
I disabled that check inside borg and now get that when trying to mount over a directory without write permissions:
```
fusermount: user has no write access to mountpoint /home/user/w/borg/mnt
```

That comes from `llfuse.init` which is not part of borg. Additionally, it is unmaintained, so discussing this issue there might not be possible. One could check if `pyfuse3` behaves the same, that successory library is maintained.

**Comment by ThomasWaldmann** [comment]
Closing this, can't do anything about it in borg.

**Comment by dakiba** [comment]
for me helped `sudo apt-get install fuse3` if someone else find this issue here.
